ADR/00xx-component-references.md

```diff
@@ -0,0 +1,147 @@
+* Introduce a new convention for PRs that lets one PR declare that it depends on another PR.
+* Important to understand: the declared reference embeddings as defined in the `ComponentReferences` and PR dependencies are two different things.
+  * `ComponentReferences` are declarations about which Components we should expect to see PR dependencies for.
+    * You say in english: “Component A references Component B.”
+  * PR dependencies can be supplied by the user without any declared `ComponentReferences` on the respective Components.
+    * You say in english: “PR #1 depends on PR #2.”
+
+### Integration-service
+
+* When [integration-service] notices a build of a *referent* Component (where another Component declares that it embeds a reference to this one), it always skips all testing (both pre-merge and post-merge). It always promotes the image to the global candidate list when the PR is merged, but it does not promote to dev environment (post-merge) and it does not create Releases (post-merge).
+  * Builds of *referent* Components never directly trigger tests, never trigger promotion, or Releases.
+* When [integration-service] notices a build of a *referer* Component (one that declares that it embeds references to others), it does testing as normal, promotes to global candidate list as normal, and it promotes to dev env as normal, and it creates Releases as normal.
+* When [integration-service] notices a build of a Component that references no other components and no other components refer to it - then it proceeds like normal, runs tests, promotes, deploys, releases, etc.
+
+### Build-service
+
+* [build-service] will propagate the digest from one Component to another as a PR – following the declared `ComponentReferences`. This is a bit like renovatebot wrapped in a controller. The [build-service] acts when it sees that a build pipelinerun has completed successfully for a *referent* image.
+  * If this is the first time it has seen this *referent* updated in this way, it will file a new pull request on the *referer* supplying the pullspec and digest of the new referent, using the same branch name as was used for the *referent* PR.
+  * If a PR on that branch already exists, update that branch with an additional commit including the new *referent* digest.
+
+The [build-service] will also update the PR it filed when the PRs that triggered it are merged or updated. It may update the description and/or it may rebase on the main branch and/or it may issue /retest. Need to figure out what we want here.
+
+### “PR dependency”
+
+TODO - describe exactly how the user specifies that one PR depends on another. PaC will have to pass this through to the PipelineRun as an annotation. For now, just assume it is possible and defined.
+
+* When [integration-service] notices a build from a PR (PR#2) that declares it depends on another PR (PR#1), it will do testing but it will also:
```

Comment:
What do you mean by "it will do testing"? Earlier, it sounded like you were saying that all testing would be skipped for the rebuilds of embedded reference.

Reply:
> When integration-service sees a build of B it **will perform testing on a Snapshot for** B as normal.

You only mention a combined snapshot with PR dependencies. While it might not be _strictly_ necessary to have a combined snapshot for only `ComponentRefernce` dependencies (the images should still be accessible for deployment regardless of what is in the snapshot and therefore deployed), should the Snapshot in this instance be merged as well for consistency?

Reply:
Hm, in the situation where you have a `ComponentReference` but you do not have a PR Dependency, then I don't think that we have enough information on hand in order to construct a merged `Snapshot`.

Note that the following should be modified to include an extra sentence on the end:

> When integration-service sees a build of A it will not perform testing on a Snapshot for A because it knows that (in order to be meaningful) it will need a Snapshot containing a new build of B. It will promote the post-merge build of A to the global candidate list even though it did not perform testing.

This ensures that, later:

> When integration-service sees a build of B it will perform testing on a Snapshot for B as normal.

.. that the Snapshot created for B can include the content from "A" which was promoted to the global candidate list. Normal behavior for integration-service is to construct Snapshots from 1) the component build that triggered it plus 2) the current contents of the global candidate list for all other components.

Reply:
With `ComponentReferences`, it is not possible to test dependencies without introducing changes to other components with PRs. Therefore, does a `ComponentReference` not trigger a PR dependency? When the former is acted upon, there is a PR opened for the user on the embedding component and the two will be dependent upon each other.

Reply:
Yeah, I think that's right. When you have a declared `ComponentReference` in place, then if you submit a PR on a _referent_, build-service should submit a PR with the digest reference update to the repository of the referer *and it should declare a PR dependency between its PR and your triggering PR*.